### PR TITLE
Add a basic PlateRead function into antha-lang

### PR DESCRIPTION
**Pull request by jake-ccl**
- Allow targets to be defined via config file
- basic antha-lang platereader PlateRead method

### CODE REVIEW

**Comment by jake-ccl** [comment]
@ddn0  Please review when you have a moment!

File: antha/anthalib/wtype/prinstruction.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package wtype
+
+import (
+	"fmt"
+)
+
+
+//  high-level instruction to a plate reader
```
Comment (ddn0) [inline antha/anthalib/wtype/prinstruction.go]: gostyle `// PRInstruction is a high-level instruction ...`

File: antha/anthalib/wtype/prinstruction.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+	ComponentOut   	   *LHComponent
+	Options            string
+}
+
+func (ins PRInstruction) String() string {
+	return fmt.Sprint("PRInstruction")
+}
+
+
+func NewPRInstruction() *PRInstruction {
```
Comment (ddn0) [inline antha/anthalib/wtype/prinstruction.go]: godoc

File: ast/ast.go [hunk 3]
```diff
@@ -234,3 +234,44 @@ func Deps(roots []Node) graph.Graph {
 		},
 	})
 }
+
+
+// FindReaching finds reachable Commands given a slice of ast.Node
```
Comment (ddn0) [inline ast/ast.go]: `// FindReachingCommands returns the set of commands that have a path to the given nodes without any intervening commands`

File: ast/ast.go [hunk 4]
```diff
@@ -234,3 +234,44 @@ func Deps(roots []Node) graph.Graph {
...
+func FindReachingCommands(nodes []Node) []Command {
+	g := ToGraph(ToGraphOpt{Roots: nodes, WhichDeps: DataDeps})
+
+	cmds := make([]Command, 0)
+	stack := make([]graph.Node, 0)
+
+	// Add immediate children to stack
+	for _, node := range nodes {
```
Comment (ddn0) [inline ast/ast.go]: gofmt this file

File: ast/ast.go [hunk 5]
```diff
@@ -234,3 +234,44 @@ func Deps(roots []Node) graph.Graph {
 		},
 	})
 }
+
+
+// FindReaching finds reachable Commands given a slice of ast.Node
+func FindReachingCommands(nodes []Node) []Command {
+	g := ToGraph(ToGraphOpt{Roots: nodes, WhichDeps: DataDeps})
+
+	cmds := make([]Command, 0)
```
Comment (ddn0) [inline ast/ast.go]: preference for `var cmds []Command` and `var stack []graph.Node`

File: ast/ast.go [hunk 6]
```diff
@@ -234,3 +234,44 @@ func Deps(roots []Node) graph.Graph {
...
+
+	// Add immediate children to stack
+	for _, node := range nodes {
+		for i := 0; i < g.NumOuts(node); i ++ {
+			stack = append(stack, g.Out(node, i))
+		}
+	}
+
+	// Breath-first search on stack
+	seen := make(map[graph.Node]bool)
```
Comment (ddn0) [inline ast/ast.go]: There is a Visit routine in `github.com/antha-lang/antha/graph` which you can use to implement this logic if you configure `VisitOpt.Visitor` or `VisitOpt.Seen` to return `ErrTraversalDone` appropriately

File: cmd/antha/cmd/run.go [hunk 7]
```diff
@@ -115,6 +115,7 @@ type runOpt struct {
 	MixerOpt               mixer.Opt
 	Drivers                []string
 	BundleFile             string
+	TargetConfigFile	   string
```
Comment (ddn0) [inline cmd/antha/cmd/run.go]: gofmt this file

File: codegen/codegen.go [hunk 8]
```diff
@@ -300,6 +305,9 @@ func (a *ir) coalesceDevices(device map[ast.Node]target.Device) {
 	a.assignment = run
 }
 
+
+
```
Comment (ddn0) [inline codegen/codegen.go]: gofmt this file

File: execute/intrinsics.go [hunk 9]
```diff
@@ -212,6 +212,47 @@ func Handle(ctx context.Context, opt HandleOpt) *wtype.LHComponent {
...
+type PlateReadOpts struct {
+	Sample  *wtype.LHComponent
+	Options string
+}
+func readPlate(ctx context.Context, opts PlateReadOpts) *commandInst {
+	inst := wtype.NewPRInstruction()
+	inst.ComponentIn = opts.Sample
+
+	// Clone the component to represent the result of the AbsorbanceRead
+	comp := inst.ComponentIn.Dup()
```
Comment (ddn0) [inline execute/intrinsics.go]: replace this with `newCompFromComp(opts.Sample)`

File: target/auto/mock.go [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+
+// MockDevice defines a mock-device
+type MockDevice struct {
+	DeviceClass	string `json:"class"`
+	DeviceName  string `json:"name"`
+	Properties  map[string]string `json:"properties"`
+}
+
```
Comment (ddn0) [inline target/auto/mock.go]: gofmt this file

File: target/auto/mock.go [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+// MockTargetConfig defines a mock-target
+type MockTargetConfig struct {
+	MockDevices []MockDevice  `json:"devices"`
+}
+
+
+// MockDevice defines a mock-device
+type MockDevice struct {
+	DeviceClass	string `json:"class"`
+	DeviceName  string `json:"name"`
```
Comment (ddn0) [inline target/auto/mock.go]: `json:"device_name"` ? `json:"device_class"`?

File: target/auto/mock.go [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+	// There was no config
+	if targetConfigFilePath == "" {
+		return nil, nil
+	}
+
+	bTargetConfig, err := ioutil.ReadFile(targetConfigFilePath)
+	if err != nil {
+		return nil, err
+	}
+	v := new(MockTargetConfig)
```
Comment (ddn0) [inline target/auto/mock.go]: traditionally written as

var v MockTargetConfig
err = yaml.Unmarshal(bTargetConfig, &v)
return &v, err
```

File: target/auto/mock.go [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+// ToDevice makes a Device from a MockDevice
+func (a *MockDevice) ToDevice() (target.Device, error) {
+	if a == nil {
+		return nil, fmt.Errorf("no device given")
+	}
+
+	// Very basic for now
+	switch a.DeviceClass {
+	case "antha_platereader_v1":
```
Comment (ddn0) [inline target/auto/mock.go]: prefer using the protouf name for clarity: `antha.platereader.v1`

File: target/platereader/platereader.go [hunk 14]
```diff
@@ -1,43 +1,162 @@
...
-func NewWOPlateReader() *PlateReader {
-	ret := &PlateReader{}
-	ret.GenericHandler = handler.GenericHandler{
-		Labels: []ast.NameValue{
-			target.DriverSelectorV1WriteOnlyPlateReader,
-		},
-		GenFunc: ret.generate,
+
+// Ensure satisfies Device interface
+var _ target.Device = &PlateReader{}
```
Comment (ddn0) [inline target/platereader/platereader.go]: save some space with:
var _ target.Device = *PlateReader(nil)

File: target/platereader/platereader.go [hunk 15]
```diff
@@ -1,43 +1,162 @@
...
+	can := ast.Request{}
+	can.Selector = append(can.Selector, target.DriverSelectorV1WriteOnlyPlateReader)
+	return can.Contains(req)
+}
+
+// MoveCost implements a Device
+func (a *PlateReader) MoveCost(from target.Device) int {
+	return 0
+}
+
```
Comment (ddn0) [inline target/platereader/platereader.go]: gofmt this file

File: target/platereader/platereader.go [hunk 16]
```diff
@@ -1,43 +1,162 @@
...
+// Compile implements a Device
+func (a *PlateReader) Compile(ctx context.Context, nodes []ast.Node) ([]target.Inst, error) {
+
+	// Find the LHComponentID for the samples to measure. We'll then search
+	// for these later.
+	lhCmpIds := make(map[string]bool)
+	for _, node := range nodes {
+		cmd, ok := node.(*ast.Command)
+		if !ok {
+			panic(fmt.Sprintf("expected *ast.Command. Got: %T", node))
```
Comment (ddn0) [inline target/platereader/platereader.go]: panic unnecessary if you use `cmd := node.(*ast.Command)` other option would be to return an error

File: target/platereader/platereader.go [hunk 17]
```diff
@@ -1,43 +1,162 @@
...
+		}
+		inst, ok := cmd.Inst.(*wtype.PRInstruction)
+		if !ok {
+			// TODO: Do we want to panic?
+			panic(fmt.Sprintf("expected PRInstruction. Got: %T", cmd.Inst))
```
Comment (ddn0) [inline target/platereader/platereader.go]: ditto; prefer returning an error over panic

File: target/platereader/platereader.go [hunk 18]
```diff
@@ -1,43 +1,162 @@
...
+	// Look for the sample locations
+	lhPlateLocations := make(map[string]string) // {cmpId: PlateId}
+	lhWellLocations := make(map[string]string) // {cmpId: A1Coord}
+	for _, cmd := range ast.FindReachingCommands(nodes) {
+		insts := cmd.Output.(*codegen.Result).Insts
+		for _, inst := range insts {
+			mix, ok := inst.(*target.Mix)
+			if !ok {
+				// TODO: Deal with other instructions
+				fmt.Printf("Expected *target.Mix, got: %T", inst)
```
Comment (ddn0) [inline target/platereader/platereader.go]: remove debugging statement

File: target/platereader/platereader.go [hunk 19]
```diff
@@ -1,43 +1,162 @@
...
+				continue
+			}
+			for _, plate := range mix.FinalProperties.Plates {
```
Comment (ddn0) [inline target/platereader/platereader.go]: refactor to a separate function

File: target/platereader/platereader.go [hunk 20]
```diff
@@ -1,43 +1,162 @@
...
+				for _, well := range plate.Wellcoords {
+					for lhCmpID := range lhCmpIds {
```
Comment (ddn0) [inline target/platereader/platereader.go]: gostyle `lhCmpIDs` or if you follow the style of other code `comp`

File: target/platereader/platereader.go [hunk 21]
```diff
@@ -1,43 +1,162 @@
...
+							lhPlateLocations[lhCmpID] = plate.ID
+							lhWellLocations[lhCmpID] = well.Crds
+						}
+					}
+				}
+			}
+		}
+	}
+
+	prInsts := make([]wtype.PRInstruction, 0)
```
Comment (ddn0) [inline target/platereader/platereader.go]:
```
var prInsts []*wtype.PRInstruction
for _, node := range nodes {
  cmd := node.(*ast.Command)
  prInsts = append(prInsts, cmd.Inst.(*wtype.PRInstruction))
```

File: target/platereader/platereader.go [hunk 22]
```diff
@@ -1,43 +1,162 @@
...
+
+// PRInstructions with the same key can be executed on the same plate-read cycle
+func pRInstructionKey(inst wtype.PRInstruction) (string, error) {
+	return fmt.Sprintf("%s", inst.Options), nil
 }
 
-func (pr *PlateReader) generate(cmd interface{}) ([]target.Inst, error) {
-	handle, ok := cmd.(*ast.HandleInst)
+// Merge PRInstructions
+func (a* PlateReader) mergePRInsts(insts []wtype.PRInstruction, wellLocs map[string]string, plateLocs map[string]string) ([]target.Inst, error) {
```
Comment (ddn0) [inline target/platereader/platereader.go]: `insts []*wtype.PRInstruction`

File: target/platereader/platereader.go [hunk 23]
```diff
@@ -1,43 +1,162 @@
...
+	insts, err := a.mergePRInsts(prInsts, lhWellLocations, lhPlateLocations)
+	if err != nil {
+		return nil, err
+	}
+	return insts, nil
+}
+
+
+// PRInstructions with the same key can be executed on the same plate-read cycle
+func pRInstructionKey(inst wtype.PRInstruction) (string, error) {
```
Comment (ddn0) [inline target/platereader/platereader.go]: *wtype.PRInstruction`

File: target/platereader/platereader.go [hunk 24]
```diff
@@ -1,43 +1,162 @@
...
+	// Merge PR instructions
+	insts, err := a.mergePRInsts(prInsts, lhWellLocations, lhPlateLocations)
+	if err != nil {
+		return nil, err
+	}
+	return insts, nil
+}
+
+
+// PRInstructions with the same key can be executed on the same plate-read cycle
```
Comment (ddn0) [inline target/platereader/platereader.go]: gostyle: `// pRInstructionKey` or better `prKey`

File: target/platereader/platereader.go [hunk 25]
```diff
@@ -1,43 +1,162 @@
...
-		Dev:   pr,
-		Label: handle.Group,
-		Calls: handle.Calls,
+	// Check for only 1 plate (for now)
+	plateLocUnique := make(map[string]bool)
+	for _, plateID := range plateLocs {
+		plateLocUnique[plateID] = true
+	}
+	if len(plateLocUnique) > 1 {
+		panic("current only supports single plate.")
```
Comment (ddn0) [inline target/platereader/platereader.go]: return error over panic

File: target/platereader/platereader.go [hunk 26]
```diff
@@ -1,43 +1,162 @@
...
+		if err != nil {
+			return nil, err
+		}
+		cmpID := inst.ComponentIn.GetID()
+		groupBy[key] = inst
+		groupedWellLocs[key] = append(groupedWellLocs[key], wellLocs[cmpID])
 	}
 
+	// Emit the driver calls
+	calls := make([]driver.Call, 0)
```
Comment (ddn0) [inline target/platereader/platereader.go]: `var calls []driver.Call`

File: target/platereader/platereader.go [hunk 27]
```diff
@@ -1,43 +1,162 @@
...
+				ProtocolOptions: inst.Options,
+			},
+			Reply: &platereader.BoolReply{},
+		}
+		calls = append(calls, call)
+	}
+
+	inst := &target.Run{
+		Dev:   a,
+		Label: "plate-Reader",
```
Comment (ddn0) [inline target/platereader/platereader.go]: `"use plate reader"`

File: antha/anthalib/wtype/prinstruction.go (hunk 1 above)
Comment (jake-ccl) [reply antha/anthalib/wtype/prinstruction.go]: done

File: ast/ast.go (hunk 3 above)
Comment (jake-ccl) [reply ast/ast.go]: Assume the algorithm is okay?

File: antha/anthalib/wtype/prinstruction.go (hunk 2 above)
Comment (jake-ccl) [reply antha/anthalib/wtype/prinstruction.go]: done

File: ast/ast.go (hunk 4 above)
Comment (jake-ccl) [reply ast/ast.go]: done

File: cmd/antha/cmd/run.go (hunk 7 above)
Comment (jake-ccl) [reply cmd/antha/cmd/run.go]: Done

File: codegen/codegen.go (hunk 8 above)
Comment (jake-ccl) [reply codegen/codegen.go]: done

File: execute/intrinsics.go (hunk 9 above)
Comment (jake-ccl) [reply execute/intrinsics.go]: done

File: target/auto/mock.go (hunk 11 above)
Comment (jake-ccl) [reply target/auto/mock.go]: done

File: target/auto/mock.go (hunk 12 above)
Comment (jake-ccl) [reply target/auto/mock.go]: done

File: target/auto/mock.go (hunk 13 above)
Comment (jake-ccl) [reply target/auto/mock.go]: done

File: target/platereader/platereader.go (hunk 14 above)
Comment (jake-ccl) [reply target/platereader/platereader.go]: var _ target.Device = (*PlateReader)(nil)

Comment (jake-ccl) [reply target/platereader/platereader.go]: done

File: target/platereader/platereader.go (hunk 15 above)
Comment (jake-ccl) [reply target/platereader/platereader.go]: done

File: target/platereader/platereader.go (hunk 16 above)
Comment (jake-ccl) [reply target/platereader/platereader.go]: done

File: target/platereader/platereader.go (hunk 17 above)
Comment (jake-ccl) [reply target/platereader/platereader.go]: done

File: target/platereader/platereader.go (hunk 18 above)
Comment (jake-ccl) [reply target/platereader/platereader.go]: done

File: target/platereader/platereader.go (hunk 20 above)
Comment (jake-ccl) [reply target/platereader/platereader.go]: done

File: target/platereader/platereader.go (hunk 21 above)
Comment (jake-ccl) [reply target/platereader/platereader.go]: done

File: target/platereader/platereader.go (hunk 22 above)
Comment (jake-ccl) [reply target/platereader/platereader.go]: done

File: target/platereader/platereader.go (hunk 19 above)
Comment (jake-ccl) [reply target/platereader/platereader.go]: done

File: ast/ast.go (hunk 6 above)
Comment (jake-ccl) [reply ast/ast.go]: Agreed to leave as it is for now, but with a comment.

**Pull request merged**